Integrate Spring Security, generate and send OTPs

liangyuanruo opened the pull request:
#13 Problem
Lack of OTP login in the application.

Addresses #10 in part.

# Solution
- Create Agency, User and OneTimePassword models
- Set up Spring Security, Spring Email
- Login controller with `POST /otp` endpoint to create, save and email OTP to user
- The rest of the login flow has been deprioritised for now.

# Improvements
- Linting configurations
- Custom checked exceptions when deviating from login flow

liangyuanruo committed:
Agency and User models

liangyuanruo committed:
UserService

liangyuanruo committed:
install spring security

liangyuanruo committed:
Add OneTimePassword class.

liangyuanruo committed:
Configure Spring Security to allow access to H2 console. Leaving all APIs open for now.

liangyuanruo committed:
Create security configuration profiles.

liangyuanruo committed:
Remove checkstyle header template

liangyuanruo committed:
Lint - fix import order

liangyuanruo committed:
Permit all login routes

liangyuanruo committed:
Implement OTP generation flow
- LoginController with POST /otp to request for OTP creation
- LoginService which verifies the Agency domain and handles the login flow
- Removed OneTimePassword dependency from User so that users can be created independently of OTPs

liangyuanruo committed:
Change login method signature to throw exception instead of returning boolean value.

liangyuanruo committed:
Use custom AgencyNotFoundException and OtpCreationException

liangyuanruo committed:
Implement OTP email flow

liangyuanruo committed:
Remove printline from OTP

liangyuanruo committed:
Avoid specifying version of spring-boot-starter-email to play nice with the rest of spring-boot

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/AbstractSecurityConfig.java:
```diff
@@ -0,0 +1,58 @@
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+
+class Profiles {
```
What's wrong with creating an enum?

liangyuanruo replied:
`@Profile` annotation takes String[] (or String) as a type

liangyuanruo replied:
it's just easier this way instead of trying to convert an enum to string

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/domain/agency/Agency.java:
```diff
@@ -0,0 +1,36 @@
+    private String emailDomain;
+
+    @SuppressWarnings("unused")
+    public Agency() {}
```
Unless my Java is rusty, there's no real need for a no-arg ctor if there's no impl

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/domain/agency/Agency.java:
```diff
@@ -0,0 +1,36 @@
+
+    @NotBlank
+    @Column(unique = true)
+    private String name;
```
no getter for the agency name? you created one for the email domain

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/domain/user/User.java:
```diff
@@ -0,0 +1,62 @@
+            return email.split("@")[1];
+        } catch (ArrayIndexOutOfBoundsException e) {
+            return null;
+        }
```
Don't do this! This isn't really exceptional!
```java
String[] emailParts = email.split("@");
return emailParts.length > 1 ? emailParts[1] : null;
```

liangyuanruo committed:
Extract configuration and profiles into a configuration package.

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/domain/otp/OneTimePassword.java:
```diff
@@ -0,0 +1,70 @@
+    private static String generateToken() throws OtpCreationException {
+        try {
+            StringBuilder sb = new StringBuilder(TOKEN_LENGTH);
+            SecureRandom sr = SecureRandom.getInstance("SHA1PRNG");
```
I'm not entirely sure this gives you the same `SecureRandom` instance, reading through the source. Given it doesn't, what you're basically getting is a non-seeded instance that will seed itself (but what with?) at the subsequent `nextInt()` call. Move this to class level field

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/domain/user/User.java:
```diff
@@ -0,0 +1,62 @@
+
+    public String getEmail() {
+        return this.email;
+    }
```
nitpick - Group getters and setters together

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/otp/OtpService.java:
```diff
@@ -0,0 +1,16 @@
+    OtpRepository otpRepository;
+
+    public OneTimePassword createOtp(String email) throws OtpCreationException {
+        return otpRepository.save(new OneTimePassword(email));
```
Did you mean to throw an `OtpCreationException` here on failure? Thinking about it.... what about using Spring's `AuthenticationServiceException` instead?

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/email/EmailService.java:
```diff
@@ -0,0 +1,45 @@
+            logger.info("OTP sent to:\t" + toEmail);
+
+        } catch (MessagingException e) {
+            throw new OtpMailException("Could not email OTP to user:\t" + toEmail, e);
```
Maybe let's genercise it to `AuthenticationServiceException`?

liangyuanruo replied:
I don't think the mail service should throw `AuthenticationServiceException` on account of separation of concerns - will let the login service throw that instead. instead i'll just throw a generic `EmailServiceException` for now.

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/login/LoginController.java:
```diff
@@ -0,0 +1,42 @@
+        try {
+            loginService.loginOrRegisterNewUser(email);
+            return new ResponseEntity<>(HttpStatus.OK);
+        } catch (Exception e) {
```
Narrow this down please..

liangyuanruo committed:
Create a mock MailSender service for non-dev profile use.

LoneRifle reviewed (commented):
Dude -
https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html

liangyuanruo committed:
Address review points
- Moved SecureRandom instance generation to class-level init with corresponding runtime exception
- Simplify custom exceptions to use Spring's AuthenticationServiceException instead
- Address other minor points

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/domain/otp/OneTimePassword.java:
```diff
@@ -0,0 +1,74 @@
+        try {
+            SR = SecureRandom.getInstance("SHA1PRNG");
+        } catch (NoSuchAlgorithmException e) {
+            throw new OtpInitRuntimeException("Could not initialise OneTimePassword", e);
```
You may want to consider throwing `IllegalStateException` instead. Also, if you prefer, consider...
```java
private static SecureRandom SR = makeSecureRandom("SHA1PRNG");

private static SecureRandom makeSecureRandom(String algorithm) {
    try {
        return SecureRandom.getInstance(algorithm);
    } catch (NoSuchAlgorithmException e) {
        throw new IllegalStateException("Could not initialise OneTimePassword", e);
    }
}
```

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/domain/otp/OneTimePassword.java:
```diff
@@ -0,0 +1,74 @@
+        return sb.toString();
+    }
+
+    private static Date calculateExpiry () {
```
We should use LocalDate instead.. it has a nicer API than the legacy Calendar package

liangyuanruo replied:
`org.hibernate.AnnotationException: @Temporal should only be set on a java.util.Date or java.util.Calendar`

liangyuanruo replied:
https://stackoverflow.com/a/49309286/2791533

liangyuanruo replied:
looks like it has to be `Date` otherwise it doesn't play well with JPA. otherwise you'll have to write your own type converters.

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/email/EmailService.java:
```diff
@@ -0,0 +1,45 @@
+            mimeMessageHelper.setSubject("One-Time Password for Rostering");
+            mimeMessageHelper.setText(
+                    "Your OTP is <b>" + otp.toString() + "</b>. " +
+                            "It will expire in 10 minutes. " +
```
Fix typo (5, not ten), and perhaps use the static constant `OneTimePassword.EXPIRY_MINUTES`

LoneRifle reviewed (changes requested):
Consider the comments. I'll rebase when I get back to work so that it no longer conflicts with master

liangyuanruo committed:
Get mock PrintScreenMailSender working for all non-production spring profiles
- Use apache commons-email for helper methods to parse MimeMessage
- Use of "!production" in profile annotation to ensure integration tests will work

liangyuanruo committed:
Improve clarity of OTP generation
- Throw IllegalStateException instead of custom runtime exception
- SecureRandom is generated from a static method call

liangyuanruo committed:
Reference EXPIRY_MINUTES in outgoing email.

liangyuanruo committed:
Refactor to use EmailServiceException to be thrown in the email service, and delegate to LoginService to throw AuthenticationServiceException.

liangyuanruo committed:
specify commons-email dependency in parent pom

liangyuanruo committed:
try running tests in dev mode

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/configuration/EmailConfig.java:
```diff
@@ -0,0 +1,106 @@
+                    .append("Attachments: ").append(p.getAttachmentList()).append("\n")
+                    .append(p.parse().getHtmlContent());
+
+            System.out.println(sb.toString());
```
```suggestion
            logger.info(sb.toString());
```

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/configuration/EmailConfig.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Override
+    public void send(MimeMessage... ms) throws MailException {
+        for (MimeMessage m: ms) {
```
nitpick - space between `m` and `:`. Elsewhere too

LoneRifle commented on employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/email/EmailService.java:
```diff
@@ -0,0 +1,45 @@
+            mimeMessageHelper.setText(
+                    "Your OTP is <b>" + otp.toString() + "</b>. " +
+                            "It will expire in " + OneTimePassword.EXPIRY_MINUTES + " minutes. " +
+                            "Please use this to login to your Rostering account.",
```
nitpick - indentation

LoneRifle reviewed (approved):
lgtm otherwise

liangyuanruo committed:
Create spring test profile

liangyuanruo committed:
address review nitpicks

liangyuanruo committed:
explicitly set mailsender bean profile to test and dev

liangyuanruo committed:
Add dummy email config to application-test properties

liangyuanruo committed:
Spring security to use dev profile during tests

liangyuanruo committed:
just use h2 db for tests
